cuda/preconditioner/jacobi_kernels.cu

```diff
@@ -97,6 +99,76 @@ __global__ void __launch_bounds__(warps_per_block *cuda_config::warp_size)
 }
 
 
+template <int max_block_size, int subwarp_size, int warps_per_block,
+          typename ValueType, typename IndexType>
+__global__ void
+__launch_bounds__(warps_per_block *cuda_config::warp_size) adaptive_generate(
+    size_type num_rows, const IndexType *__restrict__ row_ptrs,
+    const IndexType *__restrict__ col_idxs,
+    const ValueType *__restrict__ values, ValueType *__restrict__ block_data,
+    preconditioner::block_interleaved_storage_scheme<IndexType> storage_scheme,
+    precision_reduction *__restrict__ block_precisions,
+    const IndexType *__restrict__ block_ptrs, size_type num_blocks)
+{
+    // extract blocks
+    const auto block_id =
+        thread::get_subwarp_id<subwarp_size, warps_per_block>();
+    const auto block = group::this_thread_block();
+    ValueType row[max_block_size];
+    __shared__ UninitializedArray<ValueType, max_block_size * warps_per_block>
+        workspace;
+    csr::extract_transposed_diag_blocks<max_block_size, warps_per_block>(
```

Comment:
So, the matrix has to be in CSR  ? Do you see any advantage in other matrix types ?

Reply:
The matrix is first converted from whatever type it was to CSR before calling this.
To use other matrix formats we would have to implement a block-diagonal extraction algorithm for every such format.

There's no reason for not doing that, just that someone would have to do it.